src/java/org/apache/cassandra/index/SecondaryIndexManager.java

```diff
@@ -1615,4 +1673,140 @@ public static void shutdownAndWait(long timeout, TimeUnit units) throws Interrup
         shutdown(asyncExecutor, blockingExecutor);
         awaitTermination(timeout, units, asyncExecutor, blockingExecutor);
     }
+
+    /**
+     * Remove endpoints whose indexes are not queryable for the specified {@link Index.QueryPlan}.
+     *
+     * @param liveEndpoints current live endpoints where non-queryable endpoints will be removed
+     * @param keyspace to be queried
+     * @param indexQueryPlan index query plan used in the read command
+     * @param level consistency level of read command
+     */
+    public static <E extends Endpoints<E>> E filterForQuery(E liveEndpoints, Keyspace keyspace, Index.QueryPlan indexQueryPlan, ConsistencyLevel level)
+    {
+        E queryableEndpoints = liveEndpoints.filter(replica -> {
```

Comment:
I think we specifically didn't use streaming on our branches because of the allocation overhead on the critical read path. Should we do the same here?

Reply:
the `Endpoints` is part of oss c*, I think it's safer not to change it in our feature branch..we may create oss ticket if performance is an issue.. wdyt?